media-proxy/src/mesh/conn_rdma.cc

```diff
@@ -0,0 +1,388 @@
+ * and resources are cleaned up before the object is deleted.
+ *
+ * @param ctx The context used for the delete operation.
+ */
+void Rdma::on_delete(context::Context& ctx)
+{
+    // Note: this is a blocking call
+    on_shutdown(ctx);
+}
+
+/**
+ * @brief Configures the RDMA endpoint by registering the memory block with the RDMA endpoint.
+ *
+ * This function ensures that the endpoint context is initialized and the memory block for the buffer queue
+ * is allocated. It then registers it with the RDMA endpoint.
+ *
+ * @param ctx The context for the operation.
+ * @return Result::success if the endpoint is successfully configured, otherwise an appropriate error code.
+ */
+Result Rdma::configure_endpoint(context::Context& ctx)
+{
+    if (!ep_ctx) {
+        log::error("Endpoint context is not initialized");
+        return Result::error_wrong_state;
+    }
+
+    std::lock_guard<std::mutex> lock(queue_mutex);
+
+    if (!buffer_block) {
+        log::error("Memory block for buffer queue is not allocated");
+        return Result::error_out_of_memory;
+    }
+
+    // Calculate the total size of the memory block
+    size_t total_size = buffer_queue.size() * ((trx_sz + PAGE_SIZE - 1) / PAGE_SIZE * PAGE_SIZE);
+
+    // Register the entire memory block with the RDMA endpoint
+    int ret = libfabric_ep_ops.ep_reg_mr(ep_ctx, buffer_block, total_size);
+    if (ret) {
+        log::error("Memory registration failed for the buffer block")("error", fi_strerror(-ret));
+        return Result::error_memory_registration_failed;
+    }
+
+    return Result::success;
+}
+
+/**
+ * @brief Cleans up RDMA resources and resets the state.
+ *
+ * This function destroys the RDMA endpoint context if it exists, cleans up the buffer queue,
+ * and marks the RDMA as uninitialized. It ensures that all allocated resources are properly
+ * released and the state is reset to allow for reinitialization if needed.
+ *
+ * @param ctx The context used for the cleanup operation.
+ * @return Result indicating the success or failure of the cleanup operation.
+ */
+Result Rdma::cleanup_resources(context::Context& ctx)
+{
+    if (ep_ctx) {
+        int err = libfabric_ep_ops.ep_destroy(&ep_ctx);
+        if (err) {
+            log::error("Failed to destroy RDMA endpoint")("error", fi_strerror(-err));
+        }
+    }
+
+    // Clean up the buffer queue
+    cleanup_queue();
+
+    // Mark RDMA as uninitialized
+    init = false;
+
+    return Result::success;
+}
+
+/**
+ * @brief Shuts down the RDMA connection and cleans up resources.
+ *
+ * This function cancels the RDMA completion queue thread and the buffer processing thread,
+ * ensuring that they are properly joined. It also notifies buffer availability to unblock
+ * any waiting operations. If the RDMA connection was initialized, it cleans up the RDMA
+ * resources and sets the state to closed.
+ *
+ * @param ctx The context used for the shutdown operation.
+ */
+void Rdma::shutdown_rdma(context::Context& ctx)
```

Comment:
It looks like you can move content of ```Rdma::shutdown_rdma()```  into ```Rdma::on_shutdown()```

Reply:
fixed in https://github.com/OpenVisualCloud/Media-Communications-Mesh/pull/286/commits/1e7f66891946f5a904408fb568ca119b7e04eaa0